Remake custom auth UI docs

cprecioso opened the pull request:
### Description

Add specific resources for using custom auth ui in each of the login methods, in a consistent way for each one. Also added an index from the main Auth page and a note linking it from the Auth UI page.

In the process, extracted the link grid component that was duplicated in three places to create a base one that all the other use.

### Select what type of change this PR introduces:
1. [x] **Just code/docs improvement** (no functional change).
2. [ ] **Bug fix** (non-breaking change which fixes an issue).
3. [ ] **New feature** (non-breaking change which adds functionality).
4. [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected).

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+title: Create your own UI
+---
+
+The login and signup flows are pretty standard: they allow the user to sign up and then log in with their username and password. The signup flow validates the username and password and then creates a new user entity in the database.
```
Mentions `username` instead of `email`. The signup flow is more complex (the users need to verify the email for example) than the username flow.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+
+The login and signup flows are pretty standard: they allow the user to sign up and then log in with their username and password. The signup flow validates the username and password and then creates a new user entity in the database.
+
+Read more about the default username and password validation rules in the [auth overview docs](../../auth/overview.md#default-validations).
```
Mentions username instead of email.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+If you require more control in your authentication flow, you can achieve that in the following ways:
+
+1. Create your UI and use `signup` and `login` actions.
+1. Create your custom sign-up action which uses the lower-level API, along with your custom code.
```
I'd make this sentence more precise i.e. "along with your custom UI code" just to indicate that when you create the custom action, we expect you to call it from your custom UI.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+  - `password: string` <Required />
+
+  :::info
+  By default, Wasp will only save the `email` and `password` fields. If you want to add extra fields to your signup process, read about [defining extra signup fields](../auth/overview#customizing-the-signup-process).
```
Could you validate that this part is really true? I'm sure if it's possible to use extra signup fields when using the `signup` action. I think the users need to go full custom (the custom signup action).

infomiho replied:
Also, this link leads me to a 404 page.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+
+  try {
+    const providerId = createProviderId('email', args.email)
+    const providerData = await sanitizeAndSerializeProviderData({
+      hashedPassword: args.password,
+    })
```
This code is correct for the username & password, but not for email auth. If you take a look at the types, this function expects: 
```ts
export type EmailProviderData = {
    hashedPassword: string;
    isEmailVerified: boolean;
    emailVerificationSentAt: string | null;
    passwordResetSentAt: string | null;
};
```
when used for `'email'` provider. I'd test out all the code snippets written here to be 100% sure that what we are sharing with the users is correct. Otherwise, they'll get stuck and have to reach out to us in Discord.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+    await createUser(
+      providerId,
+      providerData,
+      // Any additional data you want to store on the User entity
+      {}
```
This line is where users can provide extra fields like they can when using `userSignupFields` field.

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+
+- `ensureValidEmail(args)`
+
+  Checks if the email is valid and throws an error if it's not. Read more about the validation rules [here](../auth/overview#default-validations).
```
This link leads me to 404 page

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+
+- `ensureValidPassword(args)`
+
+  Checks if the password is valid and throws an error if it's not. Read more about the validation rules [here](../auth/overview#default-validations).
```
This link leads me to 404 page

infomiho commented on web/src/components/LinkGrid.tsx:
```diff
@@ -0,0 +1,45 @@
+
+export function LinkGrid({
+  links,
+  caption = 'Click on each box for more details.',
+}: {
+  links: LinkInfo[]
+  caption?: string | false
```
This props has a weird type to me, it's optional and it can be a `string` or `false`? But it has a default value. 

For the having a default value doesn't make sense - if we already have the logic to handle a missing caption - I'd remove it. Also, I'd simplify it to be an optional `string` and that's it :)

infomiho commented on web/src/components/LinkGrid.tsx:
```diff
@@ -0,0 +1,45 @@
+  return (
+    <Link href={link} className="link-grid-box">
+      <h3>{title} »</h3>
+      {description ? <p>{description}</p> : null}
```
Can we simplify this to 
```suggestion
      {description && <p>{description}</p>}
```

infomiho commented on web/docs/auth/social-auth/create-your-own-ui.md:
```diff
@@ -0,0 +1,60 @@
+title: Create your own UI
+---
+
+The UI helpers described below are lower-level and are useful for creating your custom login links.
```
I'm kinda missing the intro we had previously:
```
[Auth UI](../../auth/ui) is a common name for all high-level auth forms that come with Wasp.

These include fully functional auto-generated login and signup forms with working social login buttons.
If you're looking for the fastest way to get your auth up and running, that's where you should look.
```

I know it's repetitive, but it might be useful context for some users. Most users that already know this will just glace over it.

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -85,6 +86,45 @@ We would also get the [Auth UI](../auth/ui) generated for us. We can set up our
 
+## Available auth UI
+
+When you have decided which auth method you want to support, you can also choose how you want to present it to your users.
```
Slight change in wording:
```suggestion
When you have decided which auth methods you want to support, you can also choose how you want to present the auth to your users.
```

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -482,7 +522,7 @@ Note that `username`s are stored in a **case-insensitive** manner.
 #### Email
 
-If you use [Email](./email) authentication, the default validations are:
+If you use [Email](./auth-ui) authentication, the default validations are:
```
Why do we link to auth ui here, when we mention email?

cprecioso replied:
(For this and all the others) I really don't know, I think some autoformatting changed this(?) I'm not sure, but I'm changing them back

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -496,7 +536,7 @@ Sometimes you want to include **extra fields** in your signup process, like firs
 
 - you need to define the fields that you want saved in the database,
-- you need to customize the `SignupForm` (in the case of [Email](./email) or [Username & Password](./username-and-pass) auth)
+- you need to customize the `SignupForm` (in the case of [Email](./auth-ui) or [Username & Password](./username-and-pass) auth)
```
Here as well, we probably don't want to link to auth UI?

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -694,7 +734,7 @@ If you are not using Wasp's Auth UI, you can skip this section. Just make sure t
 Read more about using the signup actions for:
 
-- email auth [here](../auth/email#fields-in-the-email-dict) <!-- TODO: these docs are not great at explaining using signup and login actions: https://github.com/wasp-lang/wasp/issues/1438 -->
+- email auth [here](./auth-ui#fields-in-the-email-dict) <!-- TODO: these docs are not great at explaining using signup and login actions: https://github.com/wasp-lang/wasp/issues/1438 -->
```
One more instance, do we really want to link to auth UI? The comment next to this line mentions this issue: https://github.com/wasp-lang/wasp/issues/1438 do we want to remove this comment now?

infomiho commented on web/docs/auth/username-and-pass/create-your-own-ui.md:
```diff
@@ -0,0 +1,333 @@
+---
+title: Create your own UI
```
I'd say that all the comments I left for the email auth are applicable here.

infomiho reviewed (changes requested):
Good initiative to organize the auth UI docs!

I've left some comments, but on high level:
- please make sure all the links work (no 404s)
- please make sure all the code in the docs works (test in some demo app)

cprecioso commented:
@infomiho ready for re-review

Fixed what you said, checked everything.
Further split the "custom auth" documentation: I moved everything about totally custom auth actions to "Auth > Advanced", this way it doesn't seem that it's all part of the same tutorial, and people who are looking for it can just go directly.

infomiho reviewed (changes requested):
I've gone through this once more, I've noticed again that some code examples _don't work_. I've left comments on some of the text as well, I think we could improve it a bit.

As I said earlier:

> please make sure all the code in the docs works (test in some demo

cprecioso commented:
@infomiho third time's the charm 😅

Went through all the examples in an app and made sure that they all work correctly. Made some small copy changes.

infomiho commented on web/docs/auth/advanced/custom-auth-actions.md:
```diff
@@ -0,0 +1,390 @@
+Custom sign-up actions are complex, and we don't recommend creating a custom sign-up action unless you have a good reason to do so.
+They also require you to be careful, as any small mistake will compromise the security of your app.
+
+Before using custom actions, check if our support for [custom auth UI](../overview.md#custom-auth-ui) and for [auth hooks](../auth-hooks.md) could fit well with you requirements.
```
<img width="713" alt="Screenshot 2025-04-03 at 15 08 23" src="https://github.com/user-attachments/assets/d034ba53-6b36-4ca9-add0-0266972d81a1" />

Running `npm run start` reported this warning. We should update the link to:
```suggestion
Before using custom actions, check if our support for [custom auth UI](../overview.md#custom-auth-ui) and for [auth hooks](./auth-hooks.md) could fit well with you requirements.
```

cprecioso replied:
smh that `npm run build` doesn't report it 😒

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,274 @@
+title: Create your own UI
+---
+
+In Email authentication, users log in with their email address and a password. On signup, Wasp validates the data and sends a verification email. The user record is not active until the link in the verification email is clicked. Moreover, the user can later reset their password through their email with a similar flow.
```
```suggestion
When using the email auth provider, users log in with their email address and a password. On signup, Wasp validates the data and sends a verification email. The user account is not active until the user clicks the link in the verification email. Also, the user can reset their password through a similar flow.
```

We tend to avoid passive voice as per https://wasp.sh/docs/writingguide

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,274 @@
+Read more about the default email and password validation rules in the [auth overview docs](../overview.md#default-validations).
+:::
+
+Even though Wasp offers premade [Auth UI](../ui.md) for your authentication flows, there are times where you might want more customization, so we also give you the option to create your own UI and call Wasp's auth actions on your own code, similar to how Auth UI does it under the hood.
```
```suggestion
Even though Wasp offers premade [Auth UI](../ui.md) for your authentication flows, there are times when you might want more customization, so we also give you the option to create your own UI and call Wasp's auth actions with your own code, similar to how Auth UI does it under the hood.
```
When unsure, I use Grammarly to check my writing, I'd advise you to do the same :)

infomiho commented on web/docs/auth/email/create-your-own-ui.md:
```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  if (needsConfirmation) {
+    return (
+      <p>
+        Check your email for the confirmation link. If you don't see it, check
+        spam/junk folder.
+      </p>
+    )
+  }
```
One thing I realized now, we should also then show the:
- verify email flow `import { verifyEmail } from 'wasp/client/auth'`
- reset password flow `import { requestPasswordReset, resetPassword } from 'wasp/client/auth'`

I know this adds another layer of complexity to this PR, but I think it will be _amazing_ when we have this part in the docs!

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -85,6 +86,52 @@ We would also get the [Auth UI](../auth/ui) generated for us. We can set up our
+#### Make your own UI {#custom-auth-ui}
+
+Wasp is flexible enough to let you completely customize your login and signup interface.
+We give you the login and signup functions, and you decide how and when to call them.
```
```suggestion
We give you the auth related functions, and you decide how and when to call them.
```

Since for email, we have more than login and signup functions :)

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -85,6 +86,52 @@ We would also get the [Auth UI](../auth/ui) generated for us. We can set up our
+The previously discussed options should cover the vast majority of cases. But, for the few instances where it is not enough,
+you can [create your own signup flows](./advanced/custom-auth-actions.md), with completely custom logic.
+This is not recommended, and reserved for advanced use cases.
+Please check first if other Wasp features (mainly [auth hooks](../auth-hooks.md)) can handle your requirements.
```
```suggestion
Please check first if other Wasp features (mainly [auth hooks](../auth-hooks.md)) can help you with your use case.
```

infomiho commented on web/docs/auth/advanced/custom-auth-actions.md:
```diff
@@ -0,0 +1,390 @@
+      providerData = getProviderData(existingAuthIdentity.providerData)
+      if (providerData.isEmailVerified)
+        throw new HttpError(400, 'Email already verified.')
+    }
```
I'd auto format this code, it has some wierd spacing issue 🤔

cprecioso replied:
It was already auto-formatted, but the `if` didn't have brackets so the formatting was like:

```
if (condition)
  statement
```

Anyway, it could be a bit confusing so I just added brackets around it

infomiho reviewed (commented):
Going in the right direction :)

I just now remembered that we need to also show the verify email and reset password flows for the email auth provider. Sorry for mentioning it just now. 

Left some text change suggestions + one code formatting change.

cprecioso commented:
ready @infomiho

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -85,6 +86,52 @@ We would also get the [Auth UI](../auth/ui) generated for us. We can set up our
 
 We will also have access to the `user` object in our frontend code, so we can show different UI to logged-in and logged-out users. For example, we can show the user's name in the header alongside a **logout button** or a login button if the user is not logged in.
 
+## Available auth UI
```
```suggestion
## Different ways to use auth
```

infomiho commented on web/docs/auth/overview.md:
```diff
@@ -85,6 +86,52 @@ We would also get the [Auth UI](../auth/ui) generated for us. We can set up our
+
+:::
+
+##### Custom login and signup actions (Not recommended)
```
```suggestion
#### Custom login and signup actions
```

I'd keep the size the same because it now looks like a suboption of the second option and not like the third option. 

I'd also remove the `Not recommended` - I think we explain enough with the big red banner on top to scare away inexperienced users not to use this if they don't really need to.

infomiho reviewed (approved):
Looking good, left two small recommendations

cprecioso commented:
@infomiho I solved the merge conflicts and applied your suggestions, feel free to merge if you approve
